### PR TITLE
V0.6.0+exchange declare support

### PR DESCRIPTION
**Pull request by manchicken**
## Do Not Merge!

This fixes a number of issues:
 - #46 - We need to upgrade the driver to use librabbitmq-c v0.6.0
 - #44 - We ignore arguments passed in to exchange_declare
 - #43 - Module can dish out a Math::Int64 but can't take one.

### CODE REVIEW

**Comment by manchicken** [comment]
@markwellis, @autarch , I've got this part done, I'm thinking I might spend some time next week improving test coverage.

File: RabbitMQ.xs [hunk 1]
```diff
@@ -14,9 +14,9 @@
 #define __DEBUG_ENABLED__ 0
 
 #if __DEBUG_ENABLED__ == 0
```
Comment (autarch) [inline RabbitMQ.xs]: I think that reversing this to `#if __DEBUG_ENABLED__` (true) as the first condition would be more readable.

**Comment by autarch** [comment]
I took a quick look at the diff and other than one small comment it looks good to me.

**Comment by manchicken** [comment]
What's your comment? I don't see it on the diffs.

File: RabbitMQ.xs [hunk 2]
```diff
@@ -14,9 +14,9 @@
...
-# define __DEBUG__(X) /* NOOP */
+ #define __DEBUG__(X) /* NOOP */
```
Comment (autarch) [inline RabbitMQ.xs]: I think that reversing this to `#if __DEBUG_ENABLED__` (true) as the first condition would be more readable.

**Comment by autarch** [comment]
Do you see it now?

**Comment by manchicken** [comment]
Please hold on this! I need to add more tests, I'm concerned about the portability of the numbers coming out of the Math::Int64 stuff.

**Comment by manchicken** [comment]
@markwellis , @autarch: I wrote a test which uses the web service and verifies that our 64-bit integers are going in to the server as strings. You can see this in the dumper output of the test below where the int64 types are in there as strings rather than as numbers.

I'm trying to figure out why our lovely Math::Int64 values are converting things to strings. Do you guys have any idea?

```perl
use Test::More tests => 12;
use strict;
use warnings;

# Make sure that we skip everything here if LWP::UserAgent isn't available.
if (!eval q{ use LWP::UserAgent; 1 }) {
    plan skip_all => 'LWP::UserAgent not available';
}

# Same for URI::Escape
if (!eval q{ use URI::Escape; 1 }) {
    plan skip_all => 'URI::Escape not available';
}

use Data::Dumper;
use Math::UInt64 qw/int64 uint64/;
use Sys::Hostname;
my $unique = hostname . "-$^O-$^V"; #hostname-os-perlversion
my $exchange = "nr_test_x-64bit-portability-$unique";
my $routekey = "nr_test_q-64bit-portability-$unique";
my $queue    = "nr_test_q-64bit-portability-$unique";

my $dtag1=1;

my $host = $ENV{'MQHOST'} || "dev.rabbitmq.com";

use_ok('Net::AMQP::RabbitMQ');

my $mq = Net::AMQP::RabbitMQ->new();
ok($mq);

eval { $mq->connect($host, { user => "guest", password => "guest" }); };
is($@, '', "connect");
eval { $mq->channel_open(1); };
is($@, '', "channel_open");
eval { $mq->exchange_declare(1, $exchange, { exchange_type => "fanout", passive => 0, durable => 1, auto_delete => 0 }); };
is($@, '', "exchange_declare");
my $queuename = '';
eval { $queuename = $mq->queue_declare(1, $queue, { passive => 0, durable => 1, exclusive => 0, auto_delete => 0 }); };
is($@, '', "queue_declare");
isnt($queuename, '', "queue_declare -> private name");
eval { $mq->queue_bind(1, $queue, $exchange, $routekey); };
is($@, '', "queue_bind");
eval { $mq->ack($_->{delivery_tag}) while($mq->get(1, $queuename)); };
is($@, '', "drain queue");

my $payload = "Message payload";
my $headers = {
	unsigned_integer => 12345,
	signed_integer   => -12345,
	double           => 3.141,
	string           => "string here",
     math_int64       => int64(-42 * 1_000_000_000_000),
     math_uint64      => uint64(42 * 1_000_000_000_000),
};

eval { $mq->publish(1, $routekey, $payload, { exchange => $exchange }, { headers => $headers }); };
is($@, '', "publish");
die "Fatal publish failure!" if $@;

eval { $mq->consume(1, $queuename, {consumer_tag=>'ctag', no_local=>0,no_ack=>1,exclusive=>0}); };
is($@, '', "consume");

my $rv = {};
eval { $rv = $mq->recv(); };
is($@, '', "recv");

is_deeply($rv,
          {
          'body' => $payload,
          'routing_key' => $routekey,
          'delivery_tag' => $dtag1,
          'redelivered' => 0,
          'exchange' => $exchange,
          'consumer_tag' => 'ctag',
          'props' => { 'headers' => $headers },
          }, "payload");

$mq->cancel(1, 'ctag');

# Publish a second time
eval { $mq->publish(1, $routekey, $payload, { exchange => $exchange }, { headers => $headers }); };
is($@, '', "publish");
die "Fatal publish failure!" if $@;

my $ua = LWP::UserAgent->new;

my $resp = $ua->post("http://guest:guest\@$host/mgmt/api/queues/%2F/".URI::Escape::uri_escape($queue)."/get", Content => <<"EOF");
{"count":1,"requeue":true,"encoding":"auto"}
EOF

ok $resp->is_success, "Getting message from >$queue<"
     or diag "Publishing booltest message failed: " . $resp->as_string;

diag Dumper($resp);

1;
```

**Comment by autarch** [comment]
The problem is that `amqp_kind_for_sv` doesn't know about the Math::(U)Int64 type.

**Comment by autarch** [comment]
Actually, the problem may be earlier than that. I'm trying to add checks for `SvI64OK` and `SvU64ok` to `amqp_kind_for_sv`, but those never return true. I wonder if the `SV` is somehow getting stringified somewhere else in the code.

My kingdom for more tuits. I probably won't be able to spend much more time on this today.

**Comment by manchicken** [comment]
It looks like the Math::Int64, when it's being interpreted from the C side,
is returning the value as a string. I'm going to start looking in the
direction of native 64-bit integer support.

**Comment by manchicken** [comment]
Within Math::Int64 itself

**Comment by manchicken** [comment]
I am merging this in since it's not implementing any of the things that I think could be considered not-quite-ready...

**Pull request merged**